File: packages/typescript/src/resources/templates.ts
/**
 * The `templates` resource: reusable email templates. Starter plan and up.
 * @module
 */
import type { HttpTransport } from '../http';
import { prune } from '../internal/serialize';
import type { CreateTemplateParams, Template, UpdateTemplateParams } from '../types';

/** Accessed as `axene.templates`. */
export class Templates {
  /** @internal */
  constructor(private readonly http: HttpTransport) {}

  /** List all templates, most recently updated first. */
  list(): Promise<Template[]> {
    return this.http.request<Template[]>('GET', '/v1/templates/');
  }

  /**
   * Create a template. `variables` are derived server-side from `{{name}}`
   * placeholders in the bodies, so you do not pass them.
   */
  create(params: CreateTemplateParams): Promise<Template> {
    return this.http.request<Template>(
      'POST',
      '/v1/templates/',
      prune({
        name: params.name,
        subject: params.subject,
        html_body: params.html,
        text_body: params.text,
        blocks_json: params.blocksJson,
      }),
    );
  }

  /** Fetch a single template. */
  get(id: string): Promise<Template> {
    return this.http.request<Template>('GET', `/v1/templates/${encodeURIComponent(id)}`);
  }

  /** Update a template (partial). */
  update(id: string, params: UpdateTemplateParams): Promise<Template> {
    return this.http.request<Template>(
      'PATCH',
      `/v1/templates/${encodeURIComponent(id)}`,
      prune({
        name: params.name,
        subject: params.subject,
        html_body: params.html,
        text_body: params.text,
        blocks_json: params.blocksJson,
      }),
    );
  }

  /** Delete a template. */
  delete(id: string): Promise<void> {
    return this.http.request<void>('DELETE', `/v1/templates/${encodeURIComponent(id)}`);
  }

  /** Duplicate a template (the copy's `blocks_json` is not carried over). */
  duplicate(id: string): Promise<Template> {
    return this.http.request<Template>('POST', `/v1/templates/${encodeURIComponent(id)}/duplicate`);
  }
}

File: packages/rust/src/resources/templates.rs
//! The `templates` resource: reusable email templates (Starter plan and up).

use reqwest::Method;

use super::urlencode;
use crate::error::Result;
use crate::http::HttpTransport;
use crate::models::{CreateTemplate, Template, UpdateTemplate};

/// Accessed as `client.templates()`.
#[derive(Debug, Clone)]
pub struct Templates {
    http: HttpTransport,
}

impl Templates {
    pub(crate) fn new(http: HttpTransport) -> Self {
        Self { http }
    }

    /// List all templates, most recently updated first.
    pub async fn list(&self) -> Result<Vec<Template>> {
        self.http.request(Method::GET, "/v1/templates/").await
    }

    /// Create a template. `variables` are derived server-side from `{{name}}`
    /// placeholders, so they are not passed.
    pub async fn create(&self, params: &CreateTemplate) -> Result<Template> {
        self.http
            .request_json(Method::POST, "/v1/templates/", params)
            .await
    }

    /// Fetch a single template.
    pub async fn get(&self, id: &str) -> Result<Template> {
        self.http
            .request(Method::GET, &format!("/v1/templates/{}", urlencode(id)))
            .await
    }

    /// Update a template (partial).
    pub async fn update(&self, id: &str, params: &UpdateTemplate) -> Result<Template> {
        self.http
            .request_json(
                Method::PATCH,
                &format!("/v1/templates/{}", urlencode(id)),
                params,
            )
            .await
    }

    /// Delete a template.
    pub async fn delete(&self, id: &str) -> Result<()> {
        self.http
            .request_empty(Method::DELETE, &format!("/v1/templates/{}", urlencode(id)))
            .await
    }

    /// Duplicate a template (the copy's `blocks_json` is not carried over).
    pub async fn duplicate(&self, id: &str) -> Result<Template> {
        self.http
            .request(
                Method::POST,
                &format!("/v1/templates/{}/duplicate", urlencode(id)),
            )
            .await
    }
}
